File: src/main/java/io/renren/modules/api/service/impl/TokenServiceImpl.java
package io.renren.modules.api.service.impl;

import io.renren.modules.api.dao.TokenDao;
import io.renren.modules.api.entity.TokenEntity;
import io.renren.modules.api.service.TokenService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;


@Service("tokenService")
public class TokenServiceImpl implements TokenService {
    @Autowired
    private TokenDao tokenDao;
    /**
     * 12小时后过期
     */
    private final static int EXPIRE = 3600 * 12;

    @Override
    public TokenEntity queryByUserId(Long userId) {
        return tokenDao.queryByUserId(userId);
    }

    @Override
    public TokenEntity queryByToken(String token) {
        return tokenDao.queryByToken(token);
    }

    @Override
    @Transactional(rollbackFor = Exception.class)
    public void save(TokenEntity token) throws Exception {
        tokenDao.save(token);
    }

    @Override
    @Transactional(rollbackFor = Exception.class)
    public void update(TokenEntity token) throws Exception {
        tokenDao.update(token);
    }

    @Override
    @Transactional(rollbackFor = Exception.class)
    public Map<String, Object> createToken(long userId) throws Exception {
        //生成一个token
        String token = UUID.randomUUID().toString();
        //当前时间
        LocalDate now = LocalDate.now();

        //过期时间
        LocalDateTime expireTime = LocalDateTime.now().plusHours(12);

        //判断是否生成过token
        TokenEntity tokenEntity = queryByUserId(userId);
        if (tokenEntity == null) {
            tokenEntity = new TokenEntity();
            tokenEntity.setUserId(userId);
            tokenEntity.setToken(token);
            tokenEntity.setUpdateTime(now);
            tokenEntity.setExpireTime(expireTime);

            //保存token
            save(tokenEntity);
        } else {
            tokenEntity.setToken(token);
            tokenEntity.setUpdateTime(now);
            tokenEntity.setExpireTime(expireTime);

            //更新token
            update(tokenEntity);
        }

        Map<String, Object> map = new HashMap<>(16);
        map.put("token", token);
        map.put("expire", EXPIRE);

        return map;
    }
}

File: src/main/java/io/renren/modules/api/service/impl/TokenServiceImpl.kt
package io.renren.modules.api.service.impl

import io.renren.modules.api.dao.TokenDao
import io.renren.modules.api.entity.TokenEntity
import io.renren.modules.api.service.TokenService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.time.LocalDateTime
import java.util.*


@Service("tokenService")
class TokenServiceImpl @Autowired constructor(private val tokenDao: TokenDao): TokenService {

    override fun queryByUserId(userId: Long?): TokenEntity {
        return tokenDao.queryByUserId(userId)
    }

    override fun queryByToken(token: String): TokenEntity {
        return tokenDao.queryByToken(token)
    }

    @Transactional(rollbackFor = [(Exception::class)])
    @Throws(Exception::class)
    override fun save(token: TokenEntity) {
        tokenDao.save(token)
    }

    @Transactional(rollbackFor = [(Exception::class)])
    @Throws(Exception::class)
    override fun update(token: TokenEntity) {
        tokenDao.update(token)
    }

    @Transactional(rollbackFor = [(Exception::class)])
    @Throws(Exception::class)
    override fun createToken(userId: Long): Map<String, Any> {
        //生成一个token
        val token = UUID.randomUUID().toString()
        //当前时间
        val now = LocalDate.now()

        //过期时间
        val expireTime = LocalDateTime.now().plusHours(12)

        //判断是否生成过token
        val tokenEntity: TokenEntity? = queryByUserId(userId)
        if (tokenEntity == null) {
            //保存token
            save(TokenEntity(1L,userId,token,expireTime,now))
        } else {
            tokenEntity.token = token
            tokenEntity.updateTime = now
            tokenEntity.expireTime = expireTime

            //更新token
            update(tokenEntity)
        }

        val map = HashMap<String, Any>(16)
        map.put("token", token)
        map.put("expire", EXPIRE)

        return map
    }

    companion object {
        /**
         * 12小时后过期
         */
        private val EXPIRE = 3600 * 12
    }
}
